### DISCUSSION TITLE
Improve error experience for cases that need `where T: default`

### DISCUSSION THREAD

**Discussion by ashmind** [dotnet/roslyn#52545] [Ideas] [answered]
I have ran into a recent case that needs an explicit `where T: default` -- e.g.
```csharp
#nullable enable

public interface I {
    void M<T>(T? value);
}
public class C : I {
    void I.M<T>(T? value) {}
}
```

This produces a lot of errors:
```
error CS0539: 'C.M<T>(T?)' in explicit interface declaration is not found among members of the interface that can be implemented
error CS0535: 'C' does not implement interface member 'I.M<T>(T?)'
error CS0453: The type 'T' must be a non-nullable value type in order to use it as parameter 'T' in the generic type or method 'Nullable<T>'
```

However none of them is very helpful. This is what I needed to do, for my use case:
```csharp
public class C : I {
    void I.M<T>(T? value) where T: default {}
}
```

But the errors don't hint on `where T: default` at all, and I would argue it's not a well-known syntax to know in advance.

It would be great if compiler could help more here, and suggest adding a `default` constraint as an option (probably either in `CS0453` or as a replacement for `CS0453`).

**Comment by ryzngard** [comment]
@ashmind might be worth filing this as an issue to get more traction. In general I agree with improving the messaging from the compiler for clarity though :)

**Reply by ashmind** [reply]
@ryzngard Thanks! I wanted to file it as an issue, but there is no category for this it seems:
<img src="https://user-images.githubusercontent.com/166396/114836144-74c5d480-9e26-11eb-84e9-33fd2e1eb13b.png" width="500px">

This does not seem to be a bug, and other categories do not fit either?

**Reply by Youssef1313** [reply]
There is "Open a blank issue." below.

**Reply by Youssef1313** [reply]
There is actually an existing issue. https://github.com/dotnet/roslyn/issues/48971

**Comment by RikkiGibson** [comment, marked as answer]
Issue #46458 tracks improving this experience.
